03_ROS/ghost_sim/src/v5_robot_plugin.cpp

```diff
@@ -193,68 +257,153 @@ void V5RobotPlugin::Load(gazebo::physics::ModelPtr model, sdf::ElementPtr sdf){
 	// iteration. Remove this call, the connection and the callback if not needed.
 	impl_->update_connection_ = gazebo::event::Events::ConnectWorldUpdateBegin(
 		std::bind(&V5RobotPlugin::OnUpdate, this));
+
+	// initialize joystick msg
+	impl_->sensor_msg_.joystick_msg.joystick_left_x = 0.0;
+	impl_->sensor_msg_.joystick_msg.joystick_left_y = 0.0;
+	impl_->sensor_msg_.joystick_msg.joystick_right_x = 0.0;
+	impl_->sensor_msg_.joystick_msg.joystick_right_y = 0.0;
+
+	impl_->sensor_msg_.joystick_msg.btn_a = false;
+	impl_->sensor_msg_.joystick_msg.btn_b = false;
+	impl_->sensor_msg_.joystick_msg.btn_x = false;
+	impl_->sensor_msg_.joystick_msg.btn_y = false;
+	impl_->sensor_msg_.joystick_msg.btn_up = false;
+	impl_->sensor_msg_.joystick_msg.btn_down = false;
+	impl_->sensor_msg_.joystick_msg.btn_left = false;
+	impl_->sensor_msg_.joystick_msg.btn_right = false;
+	impl_->sensor_msg_.joystick_msg.btn_l1 = false;
+	impl_->sensor_msg_.joystick_msg.btn_l2 = false;
+	impl_->sensor_msg_.joystick_msg.btn_r1 = false;
+	impl_->sensor_msg_.joystick_msg.btn_r2 = false;
+
+	impl_->joystick_msg_ = impl_->sensor_msg_.joystick_msg;
 }
 
 void V5RobotPlugin::jointToEncoderTransform(){
+	impl_->encoder_positions_ = impl_->sensor_jacobian_.completeOrthogonalDecomposition().pseudoInverse() * impl_->joint_angles_;
+	impl_->encoder_velocities_ = impl_->sensor_jacobian_.completeOrthogonalDecomposition().pseudoInverse() * impl_->joint_velocities_;
+}
+
+Eigen::VectorXd V5RobotPlugin::motorToJointTransform(Eigen::VectorXd motor_data){
+	// Actuator Jacobian 6X8 rad/s
+	// Motor Data 6X1 RPM
+	auto joint_data = impl_->actuator_jacobian_.transpose() * motor_data * impl_->RADS_TO_RPM;
+
+	return joint_data;
+}
+
+Eigen::VectorXd V5RobotPlugin::jointToMotorTransform(Eigen::VectorXd joint_data){
+	// Actuator Jacobian 6X8
+	// Joint Data 6X1
+	auto motor_data = impl_->actuator_jacobian_.inverse() * joint_data;
+	return motor_data;
+}
+
+// Wraps encoder matrix into V5Encoder state msg[21]
+void V5RobotPlugin::wrapEncoderMsg(){
 	int col_index = 0;
-	// Lamda for-each expression to get encoder values for every joint
-	for_each(begin(impl_->joint_msg_list_), end(impl_->joint_msg_list_), [&](const std::string &joint_data){
-			// Get iterator that points to corresponding joint name in joint_names_
-			auto joint_name_itr = find(impl_->joint_names_.begin(), impl_->joint_names_.end(), joint_data);
```

Comment:
is this something that could be more efficient if we had an unordered_map for looking this up from name?

Reply:
These two lines were deleted and do not exist in my branch locally?